plugins/identity/app/services/service_layer/identity_services/credential.rb

```diff
@@ -5,21 +5,70 @@ module IdentityServices
     # This module implements Openstack Credential API
     module Credential
 
-      AWSCreds = Struct.new(:access, :secret, :error)
-
-      def aws_credentials(user_id, project_id)
-        aws_creds = [] 
-        response = elektron_identity.get("users/#{user_id}/credentials/OS-EC2")
-        if !response.body["credentials"].empty?
-          creds_hash = response.body["credentials"]
-          creds_hash.each do | creds |
-            next if creds['tenant_id'] != project_id
-            aws_creds << AWSCreds.new(creds["access"] , creds["secret"], nil) 
-          end
+      def credential_map
+        @credential_map ||= class_map_proc(Identity::Credential)
+      end
+
+      def new_credential(attributes = {})
+        credential_map.call(attributes)
+      end
+
+      def find_credential!(user_id = nil, access_id = nil, options = {})
+        return nil if access_id.blank? || user_id.blank?
+        elektron_identity.get("users/#{user_id}/credentials/OS-EC2/#{access_id}", options).map_to(
+          'body.credential', &credential_map
+        )
+      end
+
+      def find_credential(user_id = nil, access_id = nil, options = {})
+        find_credential!(user_id, access_id, options)
+      rescue Elektron::Errors::ApiResponse
+        nil
+      end
+
+      # find ec2 credentials of a user for a particular project if project_id given
+      # else return all ec2 credentials of the user
+      def credentials(user_id, filter = {})
+        # caching
+        @credentials ||= elektron_identity.get("users/#{user_id}/credentials/OS-EC2", filter).map_to(
+          'body.credentials', &credential_map
+        )
+        return @credentials if filter[:tenant_id].nil?
+        @credentials.select { |credential| credential.tenant_id == filter[:tenant_id] }
+      end
+
+      def project_credentials(user_id = nil, filter = {})
+        @project_credentials = credentials(user_id, filter)
```

Comment:
Same as above. Don't hardcode credential type in the general method.

Reply:
done.